File: api/internal/dayparts/dayparts.go
// Package dayparts maps absolute slots onto the coarse day-part grid the input
// model is built around, and expands a coarse selection back into per-slot
// tiers.
//
// The grid is always evaluated in the *responder's* zone, not the organizer's.
// Two people answering the same event see different days and different day
// parts for the same instants, and each is answering about their own morning.
// Getting this wrong is invisible until someone in another country responds.
//
// Like slots and solver, this package is pure.
package dayparts

import (
	"fmt"
	"time"

	"github.com/maazinshaikh/overlap/api/internal/slots"
	"github.com/maazinshaikh/overlap/api/internal/solver"
)

// Block is one of the three coarse buckets a day is divided into.
type Block string

const (
	Morning   Block = "morning"
	Afternoon Block = "afternoon"
	Evening   Block = "evening"
)

// Boundaries between the blocks, as local wall-clock hours.
//
// The web client mirrors these constants to draw the grid. They must agree: if
// the client shows a slot under Afternoon and the server files it under
// Morning, a tap lands on a different set of slots than the one the responder
// saw. web/src/lib/dayparts.ts is the other copy.
const (
	AfternoonStartHour = 12
	EveningStartHour   = 17
)

// Valid reports whether b is a known block.
func (b Block) Valid() bool {
	switch b {
	case Morning, Afternoon, Evening:
		return true
	}
	return false
}

// ParseBlock converts the wire representation.
func ParseBlock(s string) (Block, error) {
	b := Block(s)
	if !b.Valid() {
		return "", fmt.Errorf("dayparts: unknown block %q", s)
	}
	return b, nil
}

// Cell is a coordinate on the coarse grid: one local date, one block.
type Cell struct {
	Date  slots.Date
	Block Block
}

// Of returns the cell an instant falls into when viewed from loc.
func Of(t time.Time, loc *time.Location) Cell {
	l := t.In(loc)
	return Cell{
		Date:  slots.Date{Year: l.Year(), Month: l.Month(), Day: l.Day()},
		Block: blockForHour(l.Hour()),
	}
}

func blockForHour(h int) Block {
	switch {
	case h < AfternoonStartHour:
		return Morning
	case h < EveningStartHour:
		return Afternoon
	default:
		return Evening
	}
}

// Selection is one coarse tap: a tier applied to a whole day part.
type Selection struct {
	Cell Cell
	Tier solver.Tier
}

// Group buckets slots into cells, preserving order within each cell. The web
// client uses the same grouping to lay out the grid.
func Group(starts []time.Time, loc *time.Location) map[Cell][]time.Time {
	out := make(map[Cell][]time.Time)
	for _, t := range starts {
		c := Of(t, loc)
		out[c] = append(out[c], t)
	}
	return out
}

// Expand turns coarse selections into a tier for every slot they cover.
//
// Slots in no selected cell are absent from the result rather than present with
// tier "no". The caller decides what absence means; conflating "not selected"
// with "explicitly refused" at this layer would throw away the distinction
// before anyone can use it.
//
// A later selection for the same cell wins, so a client that sends a cell twice
// gets the last value rather than an arbitrary one.
func Expand(starts []time.Time, loc *time.Location, sel []Selection) map[time.Time]solver.Tier {
	if len(sel) == 0 {
		return map[time.Time]solver.Tier{}
	}

	byCell := make(map[Cell]solver.Tier, len(sel))
	for _, s := range sel {
		byCell[s.Cell] = s.Tier
	}

	out := make(map[time.Time]solver.Tier)
	for _, t := range starts {
		if tier, ok := byCell[Of(t, loc)]; ok {
			out[t] = tier
		}
	}
	return out
}

File: web/src/lib/dayparts.ts
import type { Block, Tier } from './api';

/**
 * Boundaries between the three coarse blocks, as local wall-clock hours.
 *
 * These mirror api/internal/dayparts/dayparts.go. They must agree: the server
 * expands a coarse tap using the responder's zone, so if the client draws a
 * slot under Afternoon while the server files it under Morning, the tap lands
 * on a different set of slots than the one the responder saw.
 */
export const AFTERNOON_START_HOUR = 12;
export const EVENING_START_HOUR = 17;

export const BLOCKS: Block[] = ['morning', 'afternoon', 'evening'];

export const BLOCK_LABELS: Record<Block, string> = {
	morning: 'Morning',
	afternoon: 'Afternoon',
	evening: 'Evening'
};

export function blockForHour(hour: number): Block {
	if (hour < AFTERNOON_START_HOUR) return 'morning';
	if (hour < EVENING_START_HOUR) return 'afternoon';
	return 'evening';
}

/** A coordinate on the coarse grid: one local date, one block. */
export type Cell = { date: string; block: Block };

export const cellKey = (c: Cell) => `${c.date}|${c.block}`;

/**
 * A cache of Intl.DateTimeFormat instances, keyed by locale, zone and the
 * option set that matters here (a short tag, not the whole options object).
 *
 * Constructing a DateTimeFormat is the expensive part of formatting a date --
 * it resolves locale data and a time zone -- while calling .format() on an
 * existing instance is cheap. Every function below calls one of these on
 * every slot, and a week-long event has 100-200 slots re-formatted on every
 * grid render, so building a fresh formatter per call rather than per
 * (locale, zone, kind) turned "render the grid" into "construct several
 * hundred formatters" for no reason.
 */
const formatterCache = new Map<string, Intl.DateTimeFormat>();

function formatter(
	kind: string,
	locale: string | undefined,
	tz: string,
	options: Intl.DateTimeFormatOptions
): Intl.DateTimeFormat {
	const key = `${kind}|${locale ?? ''}|${tz}`;
	let f = formatterCache.get(key);
	if (!f) {
		f = new Intl.DateTimeFormat(locale, { ...options, timeZone: tz });
		formatterCache.set(key, f);
	}
	return f;
}

/**
 * Formats an instant as a YYYY-MM-DD local date in `tz`.
 *
 * Built from Intl parts rather than toISOString, which would give the UTC date
 * and quietly put late-evening slots on the wrong day for anyone west of
 * Greenwich.
 */
export function localDate(instant: Date, tz: string): string {
	const parts = formatter('date-ymd', 'en-CA', tz, {
		year: 'numeric',
		month: '2-digit',
		day: '2-digit'
	}).formatToParts(instant);

	const get = (t: string) => parts.find((p) => p.type === t)?.value ?? '';
	return `${get('year')}-${get('month')}-${get('day')}`;
}

/** The local hour of an instant in `tz`, 0-23. */
export function localHour(instant: Date, tz: string): number {
	const v = formatter('hour', 'en-GB', tz, { hour: '2-digit', hour12: false }).format(instant);
	// en-GB renders midnight as "24" in some engines; normalise it.
	return Number(v) % 24;
}

export function cellFor(instant: Date, tz: string): Cell {
	return { date: localDate(instant, tz), block: blockForHour(localHour(instant, tz)) };
}

export type DayGroup = {
	date: string;
	weekday: string;
	dayLabel: string;
	blocks: Record<Block, Date[]>;
};

/**
 * Groups slots into days and blocks in the viewer's zone, preserving order.
 *
 * Days with no slots at all never appear, so an event whose window is
 * weekdays-only does not render empty weekend rows.
 */
export function groupByDay(slots: Date[], tz: string): DayGroup[] {
	const byDate = new Map<string, DayGroup>();

	for (const s of slots) {
		const date = localDate(s, tz);
		let g = byDate.get(date);
		if (!g) {
			g = {
				date,
				weekday: formatter('weekday', undefined, tz, { weekday: 'short' }).format(s),
				dayLabel: formatter('day-label', undefined, tz, { month: 'short', day: 'numeric' }).format(
					s
				),
				blocks: { morning: [], afternoon: [], evening: [] }
			};
			byDate.set(date, g);
		}
		g.blocks[blockForHour(localHour(s, tz))].push(s);
	}

	return [...byDate.values()].sort((a, b) => a.date.localeCompare(b.date));
}

/** Renders a slot's start time in the viewer's zone, e.g. "9:30 am". */
export function slotLabel(instant: Date, tz: string): string {
	return formatter('slot-label', undefined, tz, { hour: 'numeric', minute: '2-digit' }).format(
		instant
	);
}

/** Tier vocabulary shared with the server. */
export const TIER_LABELS: Record<Tier, string> = {
	preferred: 'Great',
	ok: 'Works',
	if_needed: 'If needed',
	no: 'No'
};

/**
 * The coarse pass cycles through three states. Untapped means no, which is why
 * the cycle returns to undefined rather than to an explicit tier.
 */
export const COARSE_CYCLE: (Tier | undefined)[] = [undefined, 'ok', 'if_needed'];

export function nextCoarse(current: Tier | undefined): Tier | undefined {
	const i = COARSE_CYCLE.findIndex((t) => t === current);
	return COARSE_CYCLE[(i + 1) % COARSE_CYCLE.length];
}

/**
 * The fine pass cycles all four tiers. It is the only place `preferred` can be
 * expressed, which is what makes the preference tiers more than decoration.
 */
export const FINE_CYCLE: Tier[] = ['preferred', 'ok', 'if_needed', 'no'];

export function nextFine(current: Tier): Tier {
	const i = FINE_CYCLE.indexOf(current);
	return FINE_CYCLE[(i + 1) % FINE_CYCLE.length];
}
